clang-tools-extra/clang-tidy/cppcoreguidelines/OwningMemoryCheck.cpp

```diff
@@ -147,10 +161,52 @@ void OwningMemoryCheck::registerMatchers(MatchFinder *Finder) {
   // Matching on functions, that return an owner/resource, but don't declare
   // their return type as owner.
   Finder->addMatcher(
-      functionDecl(hasDescendant(returnStmt(hasReturnValue(ConsideredOwner))
-                                     .bind("bad_owner_return")),
-                   unless(returns(qualType(hasDeclaration(OwnerDecl)))))
-          .bind("function_decl"),
+      functionDecl(
+          decl().bind("function_decl"),
+          hasBody(stmt(
+              stmt().bind("body"),
+              hasDescendant(
+                  returnStmt(hasReturnValue(ConsideredOwner),
+                             // Ignore sub-lambda expressions
+                             hasAncestor(stmt(anyOf(equalsBoundNode("body"),
```

Comment:
Why here we need to check `hasAncestor(stmt(anyOf(equalsBoundNode("body")`, we get this `returnStmt` from `body hasDescendant`.

Reply:
there is hasAncestor body or lambda. Simply it will match an first one that it finds, and that we bind to 'scope'.
In next lines we simply check if body and scope is same, if not then we in lambda.